### PR TITLE
Struct and union declarations

### PR DESCRIPTION
**Pull request by johnthagen**
#30

### CODE REVIEW

**Comment by johnthagen** [comment]
@nrc I've found one wrinkle so far.

The second case of a brace that doesn't fit within the left margin:

```rust
struct Fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo {
    a: A,
    b: B,
}
```

Errors for `rustfmt`:

```bash
> cargo +nightly fmt -- --write-mode diff
Rustfmt failed at main.rs:1: line exceeded maximum length
 (maximum: 100, found: 101) (sorry)
```

Looks like https://github.com/rust-lang-nursery/rustfmt/issues/1080

**Comment by joshtriplett** [comment]
This looks reasonable to me.

**Comment by johnthagen** [comment]
I do wonder if the first special case:

```rust
struct Foo /* Margin ends at `Foo` */
{
    a: A,
    b: B,
}
```

is worth allowing. It's not supported by `rustfmt`, and it seems pretty unlikely that it would be run into often. There's only two characters this saves you (the `_{`) to make the struct name longer before you have to restructure your code anyway.

Are there times when structs would be *defined* starting with a lot of rightward drift? I can only think of nested `mod`s.

**Comment by joshtriplett** [comment]
@johnthagen Technically you can define a structure inside a function or block, so it can start arbitrarily far to the right.  In practice, I agree that that special case doesn't seem worth dealing with.  If `struct Name {` passes the right margin, let it.

**Comment by nrc** [comment]
> There's only two characters this saves you

Yeah, I agree this is not worth the effort.

We could either leave it undefined (we have to at some point), or we could allow a linebreak between `struct` and `Foo`. I'm easy either way, tbh. My current feeling is that once we get into this level of edge case, it should be left for the formatter to decide. I suppose I think it is better to break the generics or whatever than to break in that position.

File: text/0001-struct-union-declaration.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    long_name: 
+        LongName,
+}
+```
+
+
+# Implementation
+[implementation]: #implementation
+
+
```
Comment (nrc) [inline text/0001-struct-union-declaration.md]: You should just comment that this is basically done already

File: text/0001-struct-union-declaration.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+# Rationale
+[rationale]: #rationale
+
+Requiring a trailing comma on the last struct field has the following benefits:
+
+- Avoids the need to add a comma when adding another field
+- Avoids the need to delete a comma with deleting a field
+- Allows easier copying of fields between different structs
+
```
Comment (nrc) [inline text/0001-struct-union-declaration.md]: I would say that it follows our general principle to have trailing commas and reference the issue where that is discussed

File: text/0001-struct-union-declaration.md (hunk 1 above)
Comment (nrc) [inline text/0001-struct-union-declaration.md]: And mention not done for unions, referencing https://github.com/rust-lang-nursery/rustfmt/issues/1157

File: text/0001-struct-union-declaration.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+[alternatives]: #alternatives
+
+Some prefer to omit the trailing comma from the final struct field.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Unions have not been stabilized yet, and thus the syntax could possibly change. `rustfmt` does not
+currently have testing for union declarations.
```
Comment (nrc) [inline text/0001-struct-union-declaration.md]: Could you add unions examples alongside some of the structs ones and highlight any differences (I don't think there actually are any though).

**Comment by johnthagen** [comment]
> We could either leave it undefined (we have to at some point), or we could allow a linebreak between `struct` and `Foo`

@nrc I took out this special case out as I agree with @joshtriplett that it's just not worth the effort.

I believe I touched on your other comments.

File: text/0001-struct-union-declaration.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+Unions are currently not supported by `rustfmt:` 
+[rust-lang-nursery/rustfmt#1157](https://github.com/rust-lang-nursery/rustfmt/issues/1157)
+
+
+# Rationale
+[rationale]: #rationale
+
+Requiring a trailing comma on the last struct field has the following benefits:
+
+- Follows the principle to use a trailing comma when it followed by newline:
```
Comment (VictorGavrish) [inline text/0001-struct-union-declaration.md]: Typo: "when followed", "when it is followed"?

Comment (johnthagen) [inline text/0001-struct-union-declaration.md]: Nice catch, should be fixed now (along with this missing "**a** newline") in the same line.

**Comment by nrc** [comment]
Now entering FCP!

**Pull request merged**